### PR TITLE
Use the executor for the part that can be parallell when rebuilding i…

### PR DESCRIPTION
**Pull request by baldersheim**
…ndex on load.

@toregge PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp [hunk 1]
```diff
@@ -220,23 +223,45 @@ DenseTensorAttribute::onLoad(vespalib::Executor *executor)
...
             tensorReader.readTensor(raw.data, _denseTensorStore.getBufSize());
             _refVector.push_back(raw.ref);
             if (_index && !use_index_file) {
-                // This ensures that get_vector() (via getTensor()) is able to find the newly added tensor.
-                setCommittedDocIdLimit(lid + 1);
-                _index->add_document(lid);
-                if ((lid % 256) == 0) {
-                    commit();
+                if (executor != nullptr) {
+                    while (pending > 1000) { std::this_thread::yield(); }
```
Comment (toregge) [inline searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:237]: Just yield causes wasted cpu usage on systems with many CPU cores. Consider using a condition variable.

**Review by toregge — APPROVED** [review]

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp]: Refactored and fixed. PTAL

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp [hunk 2]
```diff
@@ -204,11 +204,84 @@ DenseTensorAttribute::extract_cells_ref(DocId docId) const
...
+          _pending(0)
+    {}
+    void load(uint32_t lid, vespalib::datastore::EntryRef ref) override {
+        {
+            std::unique_lock guard(_mutex);
+            while (_pending > MAX_PENDING) {
+                _cond.wait(guard);
+            }
+        }
+        while (_pending > 1000) { std::this_thread::yield(); }
```
Comment (toregge) [inline searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:232]: Leftover code that should be removed ?

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp [hunk 3]
```diff
@@ -204,11 +204,84 @@ DenseTensorAttribute::extract_cells_ref(DocId docId) const
...
+        ++_pending;
+        _shared_executor.execute(vespalib::makeLambdaTask([this, ref, lid]() {
+            auto prepared = _attr._index->prepare_add_document(lid, _attr._denseTensorStore.get_typed_cells(ref), _attr.getGenerationHandler().takeGuard());
+            _complete_executor.execute(vespalib::makeLambdaTask([this, lid, result=std::move(prepared)]() mutable {
```
Comment (toregge) [inline searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:236]: Note that this allows "out of order" completions. If we want "in order" completions (as during feeding) then the shared executor should populate a promise and the complete executor task should take a future as argument and be scheduled in the foreground (not by the shared executor task).

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp (hunk 2 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp]: Fixed

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp (hunk 3 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp]: Out of order is not a problem, as we do not have any ordering guarantees in vespa either. As ANN is in nature approximative that should be even less of an issue.
However I rewrote to use a queue and let the foreground thread do the completion too.
That avoids creating an extra thread and any issues with 2 threads modifying the attribute, even if they did seem to be disjunct.

**Comment by baldersheim** [comment]
@toregge PTAL

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp [hunk 4]
```diff
@@ -213,57 +213,96 @@ class Loader {
...
-            }));
-        }));
     }
-    void wait_complete() override {
-        std::unique_lock guard(_mutex);
-        while (_pending > 0) {
-            _cond.wait(guard);
+
+    void complete(uint32_t lid, std::unique_ptr<PrepareResult> prepared) {
+        _attr.setCommittedDocIdLimit(std::max(_attr.getCommittedDocIdLimit(), lid + 1));
```
Comment (toregge) [inline searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:257]: This assumes in-order completion. The documents with highest lid might not be visible after load due to a later
complete of a document with a lower lid when using out-of-order completions.

**Review by toregge — APPROVED** [review]

**Pull request merged**